apps/rush-lib/src/logic/pnpm/PnpmProjectShrinkwrapFile.ts

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as crypto from 'crypto';
+import { InternalError, JsonFile } from '@rushstack/node-core-library';
+
+import { BaseProjectShrinkwrapFile } from '../base/BaseProjectShrinkwrapFile';
+import {
+  PnpmShrinkwrapFile,
+  IPnpmShrinkwrapDependencyYaml,
+  IPnpmShrinkwrapImporterYaml
+} from './PnpmShrinkwrapFile';
+import { DependencySpecifier } from '../DependencySpecifier';
+import { RushConstants } from '../RushConstants';
+
+/**
+ *
+ */
+export class PnpmProjectShrinkwrapFile extends BaseProjectShrinkwrapFile {
+  /**
+   * Generate and write the project shrinkwrap file to <project>/.rush/temp/shrinkwrap-deps.json.
+   * @returns True if the project shrinkwrap was created or updated, false otherwise.
+   */
+  public async updateProjectShrinkwrapAsync(): Promise<void> {
+    // If the feature is not enabled, clean up the shrinkwrap and return
+    if (
+      this.project.rushConfiguration.experimentsConfiguration.configuration
+        .legacyIncrementalBuildDependencyDetection
+    ) {
+      return this.deleteIfExistsAsync();
+    }
+
+    const projectShrinkwrapMap: Map<string, string> | undefined = this.shrinkwrapFile.isWorkspaceCompatible()
+      ? this.generateWorkspaceProjectShrinkwrapMap()
+      : this.generateLegacyProjectShrinkwrapMap();
+
+    return projectShrinkwrapMap ? this.saveAsync(projectShrinkwrapMap) : this.deleteIfExistsAsync();
+  }
+
+  protected generateWorkspaceProjectShrinkwrapMap(): Map<string, string> | undefined {
+    // Obtain the workspace importer from the shrinkwrap, which lists resolved dependencies
+    const importerKey: string = this.shrinkwrapFile.getImporterKeyByPath(
+      this.project.rushConfiguration.commonTempFolder,
+      this.project.projectFolder
+    );
+    const importer: IPnpmShrinkwrapImporterYaml | undefined = this.shrinkwrapFile.getImporter(importerKey);
+    if (!importer) {
+      // It's not in here. This is possible when perfoming filtered installs
+      return undefined;
```

Comment:
Should this check if this is a project that wasn't installed? And/or, should we never try to write the per-project shrinkwrap for projects that aren't in the installation set?

Reply:
Talked offline, but basically we should be getting out of the business of validating the lockfile and throwing when it doesn't match what we expect, mostly because we keep getting bitten by this. By returning `undefined` here, it gives us the ability to quietly fail, and then we delete the shrinkwrap for projects not in the installation set in order to clean up the current state.